Add examples of route editing

cathay4t opened the pull request.

tyll commented:
Thank you! Please add tests for the examples to https://github.com/nmstate/nmstate/blob/master/tests/integration/examples_test.py

cathay4t commented:
@tyll New patch uploaded with integration test on example files.

tyll commented on tests/integration/examples_test.py:
```diff
@@ -49,3 +50,23 @@ def test_add_remove_linux_bridge(eth1_up):
+    Test adding, downing and removing a vlan
+    """
+    route = {
+        Route.DESTINATION: '192.0.3.0/24',
```
Not a reserverd IP range for tests.

tyll commented on examples/eth1_del_route.yml:
```diff
@@ -0,0 +1,16 @@
+
+routes:
+  config:
+  - next-hop-interface: eth1
```
IMHO this should at least specify some more criteria for the specific route or the file should be called `eth1_delete_all_routes.yml` to indicate that it operates on all routes.

tyll commented on tests/integration/examples_test.py:
```diff
@@ -49,3 +50,23 @@ def test_add_remove_linux_bridge(eth1_up):
+                       cleanup='eth1_del_route.yml') as desired_state:
+        assertlib.assert_state(desired_state)
+        with example_state('eth1_del_route.yml') as desired_state:
+            assertlib.assert_route_absent(route)
```
Since the example removes all routes, the assertion should IMHO assert that there are no routes for eth1 left in config. Otherwise it would not catch if for example for some reason the metric is just changed slightly.

cathay4t commented:
@EdDev New patch uploaded.

Changes:
 * Use ipv4 test network in example.
 * Renamed the example file to `eth1_del_all_routes.yml`.
 * Changed the test assert function to make sure no routes remained after route deletion.

EdDev commented on tests/integration/examples_test.py:
```diff
@@ -70,3 +71,23 @@ def test_dns_edit(eth1_up):
+    with example_state('eth1_add_route.yml',
+                       cleanup='eth1_del_all_routes.yml') as desired_state:
+        assertlib.assert_state(desired_state)
+        with example_state('eth1_del_all_routes.yml') as desired_state:
```
Why do you need these two lines? Upon the previous contextmanager exit, the `cleanup` is called which does the exact thing.

cathay4t replied:
Removed. Thanks.

cathay4t replied:
BTW. Do you why we do it for `test_add_down_remove_vlan`:

```python
    with example_state(
        'vlan101_eth1_up.yml', cleanup='vlan101_eth1_absent.yml'
    ) as desired_state:
        assertlib.assert_state(desired_state)
        with example_state('vlan101_eth1_down.yml') as desired_state:
            assertlib.assert_absent(vlan_ifname)

    assertlib.assert_absent(vlan_ifname)
```

EdDev replied:
One is `down` and the other is `absent`.

EdDev commented on tests/integration/examples_test.py:
```diff
@@ -70,3 +71,23 @@ def test_dns_edit(eth1_up):
+    """
+    Test adding, downing and removing a vlan
+    """
+    route = {
```
This is a leftover?

EdDev replied:
This was not resolved.

cathay4t commented:
New patch uploaded.

 Changes:
 * Removed the un-needed two check on route removal.
 * Fixed the incorrect function comments.

EdDev commented on examples/eth1_del_all_routes.yml:
```diff
@@ -0,0 +1,16 @@
+---
+interfaces:
```
Do we need this section?
If it is dropped, we can show that routes may be removed without specifying the interface itself.

cathay4t replied:
We need it as we don't support editing routes without interface yet.

https://nmstate.atlassian.net/browse/NMSTATE-212

EdDev replied:
Ohh, right. Is specifying the name enough?

cathay4t replied:
Changed to include name only.
